app/models/podcast_import.rb

```diff
@@ -1,379 +1,61 @@
-    default_feed.feed_images.reset
-
-    default_feed.copy_media
-  end
-
-  def build_podcast_attributes
-    podcast_attributes = {}
-
-    %w[copyright language update_frequency update_period].each do |atr|
-      podcast_attributes[atr.to_sym] = clean_string(feed.send(atr))
-    end
-
-    podcast_attributes[:summary] = clean_text(feed.itunes_summary)
-    podcast_attributes[:link] = clean_string(feed.url)
-    podcast_attributes[:explicit] = explicit(feed.itunes_explicit, "false")
-    podcast_attributes[:new_feed_url] = clean_string(feed.itunes_new_feed_url)
-    podcast_attributes[:enclosure_prefix] ||= enclosure_prefix(feed.entries.first)
-    podcast_attributes[:feedburner_url] ||= feedburner_url(feed.feedburner_name)
-    podcast_attributes[:url] ||= feedburner_url(feed.feedburner_name)
-
-    podcast_attributes[:author] = person(feed.itunes_author)
-    podcast_attributes[:managing_editor] = person(feed.managing_editor)
-
-    owner = owner(feed.itunes_owners)
-    podcast_attributes[:owner_name] = owner[:name]
-    podcast_attributes[:owner_email] = owner[:email]
-
-    podcast_attributes[:categories] = parse_categories(feed)
-    podcast_attributes[:complete] = (clean_string(feed.itunes_complete) == "yes")
-    podcast_attributes[:copyright] ||= clean_string(feed.media_copyright)
-    podcast_attributes[:keywords] = parse_keywords(feed)
-    podcast_attributes[:serial_order] = feed.itunes_type && !!feed.itunes_type.match(/serial/i)
-    podcast_attributes[:locked] = true # won't publish feed until this is set to false
-
-    podcast_attributes[:title] = clean_string(feed.title)
-    podcast_attributes[:subtitle] = clean_string(podcast_short_desc(feed))
-    podcast_attributes[:description] = feed_description(feed)
-
-    podcast_attributes
-  end
-
-  def create_or_update_podcast!
-    if config[:episodes_only]
-      raise "No podcast for import of episodes only" unless podcast&.persisted?
-      raise "No podcast for import of episodes only" if !podcast
-      return podcast
-    end
-
-    self.podcast ||= Podcast.new
-    podcast.assign_attributes(**build_podcast_attributes)
-    update!(podcast: podcast)
-
-    update_itunes_categories(feed)
-    update_images(feed)
-
-    podcast
-  end
-
-  def enclosure_prefix(podcast_item)
-    prefix = ""
-    link = [podcast_item.feedburner_orig_enclosure_link,
-      podcast_item.enclosure.try(:url),
-      podcast_item.media_contents.first.try(:url)].find do |url|
-      url.try(:match, /podtrac/) || url.try(:match, /blubrry/)
-    end
-    if (scheme = link.try(:match, /^https?:\/\//))
-      prefix += scheme.to_s
-    end
-    if (podtrac = link.try(:match, /\/(\w+\.podtrac.com\/.+?\.mp3\/)/))
-      prefix += podtrac[1]
-    end
-    if (blubrry = link.try(:match, /\/(media\.blubrry\.com\/[^\/]+\/)/))
-      prefix += blubrry[1]
-    end
-    prefix
-  end
+  def status_from_episodes!
+    stats = with_lock { episode_imports.group(:status).count }
```

Comment:
what does `lock`ing do in this context?

Reply:
It puts a row lock on this single `podcast_import.id`.  So only one thread/process can run that "count by status" SQL query at a time.

Imagine there are only 2 episode imports, and this method is called at the exact same time from both of them.  So the query returns `{"importing": 1, "complete": 1}` to both, and neither completes this PodcastImport.

In hindsight - I should probably wrap this whole method in that `with_lock do`.  To prevent the podcast_import status update from similarly racing.  I'll make that update.